### ISSUE TITLE
ZipDecoder fails to read its own zip file with german "umlauts" in filenames

### ISSUE THREAD

**Issue by grundid** [brendan-duncan/archive#72]
When I try to encode a zip file with filenames that contain german umlauts (öäüÖÄÜß) the encoder does not complain and creates a zip file. However the ZipDecoder cannot decode the file and fails with the following exception. Also many other unzipping programs fail to decode such an archive.

```
FormatException: Bad UTF-8 encoding 0xf6 (at offset 4)
dart:convert                                        Utf8Decoder.convert
package:archive/src/util/input_stream.dart 183:25   InputStream.readString
package:archive/src/zip/zip_file_header.dart 44:26  new ZipFileHeader
package:archive/src/zip/zip_directory.dart 53:23    new ZipDirectory.read
package:archive/src/zip_decoder.dart 21:30          ZipDecoder.decodeBuffer
package:archive/src/zip_decoder.dart 15:12          ZipDecoder.decodeBytes
test/zip_file_names_test.dart 22:38                 main.<fn>

```

Here is a test code to reproduce this:

```dart
test('file encoding zip file', () {
    String origialFileName = "fileöäüÖÄÜß.txt";
    List<int> bytes = Utf8Codec().encode("test");
    ArchiveFile archiveFile = ArchiveFile(origialFileName, bytes.length, bytes);
    Archive archive = Archive();
    archive.addFile(archiveFile);

    ZipEncoder encoder = ZipEncoder();
    ZipDecoder decoder = ZipDecoder();

    var encodedBytes = encoder.encode(archive);

    Archive archiveDecoded = decoder.decodeBytes(encodedBytes);

    ArchiveFile decodedFile = archiveDecoded.files.first;

    expect(decodedFile.name, origialFileName);
  });
```

My workaround right now is to encode the filename to utf8 before it is passed to the ArchiveFile like this:

```dart
    String utf8EncodedFileName =
        String.fromCharCodes(Utf8Encoder().convert(origialFileName));

```

I think this should happen in the encoder.

**Comment by brendan-duncan** [comment]
Thanks for the example, I'll take a look at it

**Comment by brendan-duncan** [comment]
I pushed the fix, and will publish a new version shortly.

**Comment by grundid** [comment]
This has been fixed in 2.0.10. Thanks!
